app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -0,0 +1,148 @@
+
+import android.app.Application
+import android.app.Application.BIND_AUTO_CREATE
+import android.content.BroadcastReceiver
+import android.content.ComponentName
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import android.content.ServiceConnection
+import android.os.IBinder
+import androidx.core.content.ContextCompat.startForegroundService
+import net.freehaven.tor.control.TorControlCommands.EVENT_CIRCUIT_STATUS
+import net.freehaven.tor.control.TorControlCommands.EVENT_ERR_MSG
+import net.freehaven.tor.control.TorControlCommands.EVENT_HS_DESC
+import net.freehaven.tor.control.TorControlCommands.EVENT_NEW_DESC
+import net.freehaven.tor.control.TorControlCommands.EVENT_OR_CONN_STATUS
+import net.freehaven.tor.control.TorControlCommands.EVENT_WARN_MSG
+import net.freehaven.tor.control.TorControlCommands.HS_ADDRESS
+import net.freehaven.tor.control.TorControlConnection
+import org.slf4j.LoggerFactory.getLogger
+import org.torproject.jni.TorService
+import org.torproject.jni.TorService.ACTION_STATUS
+import org.torproject.jni.TorService.EXTRA_STATUS
+import java.io.IOException
+import java.util.Collections
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.coroutines.Continuation
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+private val LOG = getLogger(TorManager::class.java)
+private val EVENTS = listOf(
+    EVENT_CIRCUIT_STATUS, // this one is needed for TorService to function
+    EVENT_OR_CONN_STATUS,
+    EVENT_HS_DESC,
+    EVENT_NEW_DESC,
+    EVENT_WARN_MSG,
+    EVENT_ERR_MSG,
+)
+
+@Singleton
+class TorManager @Inject constructor(
+    private val app: Application,
+) {
+    private var torService: OnionService? = null
+    private var broadcastReceiver: BroadcastReceiver? = null
+    private val serviceConnection = object : ServiceConnection {
+        override fun onServiceConnected(className: ComponentName, service: IBinder) {
+            LOG.info("OnionService connected")
+            val binder = service as OnionService.OnionBinder
+            torService = binder.service
+        }
+
+        override fun onServiceDisconnected(arg0: ComponentName) {
+            LOG.info("OnionService disconnected")
+            torService = null
+        }
+    }
+
+    /**
+     * Starts [TorService] and creates a new onion service.
+     * Suspends until the address of the onion service is available.
+     */
+    suspend fun start(port: Int): String = suspendCoroutine { continuation ->
+        broadcastReceiver = object : BroadcastReceiver() {
+            override fun onReceive(context: Context, i: Intent) {
+                when (i.getStringExtra(EXTRA_STATUS)) {
+                    TorService.STATUS_STARTING -> LOG.debug("TorService: Starting...")
+                    TorService.STATUS_ON -> {
+                        // we are receiving a broadcast from the service, it must be set
```

Comment:
Is it possible for onServiceDisconnected() to be called concurrently with the system delivering this broadcast from the TorService to this receiver, so that `torService` ends up being null again at this point?

Reply:
As far as I can see both calls would be made from the same main UI thread, so can't be done concurrently. However, if we move to running OnionService in a separate process, we won't keep a reference to it anymore, so the point is then moot.

Reply:
:+1: